File: 프로그래머스/2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.py
import math

def convert(n: int, k: int) -> str:
    if k < 2:
        raise ValueError("k must be >= 2")  # k==1 무한루프 방지
    if n == 0:
        return "0"
    res = []
    while n > 0:
        res.append(str(n % k))
        n //= k
    return ''.join(reversed(res))

def is_prime(n: int) -> bool:
    if n < 2:
        return False
    if n % 2 == 0:
        return n == 2
    r = int(math.isqrt(n))
    for i in range(3, r + 1, 2):
        if n % i == 0:
            return False
    return True

def solution(n: int, k: int) -> int:
    s = convert(n, k)
    ans = 0
    cur = []
    for ch in s:
        if ch == '0':
            if cur:
                val = int(''.join(cur))
                if is_prime(val):
                    ans += 1
                cur.clear()
        else:
            cur.append(ch)
    if cur:
        if is_prime(int(''.join(cur))):
            ans += 1
    return ans

File: 프로그래머스/2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using pi = pair<int, int>;
using pll = pair<ll, ll>;

string Convert(int n, int k)
{
    if (n == 0) return "0";
    string res;
    while (n > 0)
    {
        const auto d = n % k;
        res.push_back((char)d + '0');
        n /= k;
    }
    reverse(res.begin(), res.end());
    return res;
}
bool IsPrime(unsigned long long n)
{
    if (n < 2) return false;
    if ((n & 1ULL) == 0ULL) return n == 2ULL;
    for (unsigned long long i = 3; i <= n / i; i += 2)
    {
        if (n % i == 0ULL) return false;
    }
    return true;
}
int solution(int n, int k)
{
    const string target = Convert(n, k);
    string stack; stack.reserve(target.size());
    size_t ans = 0;
    for (const auto ch : target)
    {
        if (ch == '0')
        {
            if (!stack.empty() && IsPrime(stoull(stack)))
            {
                ++ans;
            }
            stack.clear();
        }
        else
        {
            stack.push_back(ch);
        }
    }
    if (!stack.empty())
    {
        ans += IsPrime(stoull(stack));
    }
    return ans;
}